ansible/inventories/gateways/group_vars/all.yml

```diff
@@ -0,0 +1,24 @@
+---
+ansible_connection: ssh
+ansible_ssh_user: root
+deploy_user: root
+
+go_ipfs_version: "v0.4.22"
+
+go_ipfs_env:
```

Comment:
Hmm there's something I don't understand, is this the only configuration changes to the defaults that the gateways use?

And `go_ipfs_systemd_env` should not also include these? Should not ipfs running in docker pretty much the same config as ipfs running as binary?

Also, I think we should also use stable peer IDs for gateways (at least for some of them).

I am inclined to say that we should template the configuration directly and forget about env vars. It is much easier to silently screw an env var than the json configuration, and changes in defaults will affect gateway configs.

Also, not sure if we need to keep a dockerized way of running around if running the binary directly is the preferred way (with reason). Docker is not adding anything useful to the mix.

Reply:
> Hmm there's something I don't understand, is this the only configuration changes to the defaults that the gateways use?

we're not running go-ipfs as a container on the gateways yet so this config stanza is irrelevant at the moment, `go_ipfs_systemd_env` is what sets the actual config in the SystemD unit. will remove `go_ipfs_env` for now.
> 


> Also, I think we should also use stable peer IDs for gateways (at least for some of them).

This is currently the case when deploying via these scripts. We only get new IDs when re-deploying via Terraform. 
>

> I am inclined to say that we should template the configuration directly and forget about env vars. It is much easier to silently screw an env var than the json configuration, and changes in defaults will affect gateway configs.

The ENV vars get injected into `config.json` via the `/usr/local/bin/start_ipfs` ENTRYPOINT script, so if an ENV var is screwed up somehow then go-ipfs will fail to de-serialize its config JSON and crash on startup. 
I'm not a huge fan of templates for this particular case because it can be burdensome to maintain the template as config keys get added/removed. With ENVs, we only affect the settings we want to change.
>

> Also, not sure if we need to keep a dockerized way of running around if running the binary directly is the preferred way (with reason). Docker is not adding anything useful to the mix.

One benefit of running go-ipfs as a Docker container is that we can scrape the Docker daemon for individual container metrics like CPU, I/O,restarts etc similar to the ones available via `docker stats` and `docker inspect` as well as the image tag which makes it easier to track what version of go-ipfs each instance is running. It's also easier to set resource limits and it enables writing health/readiness checks for each container. 
Another (less exciting) benefit would be that we can upload the Docker image to a repo only once and then download it from each instance instead of (currently) pushing the binary via SSH (through Ansible). 
>

Reply:
> 

I'm not sure where the injection happens. Definitely not in `start_ipfs` -> https://github.com/ipfs/go-ipfs/blob/master/bin/container_daemon ?

Reply:
ah sorry, we have a custom docker build. I'm not sure we should maintain a custom build either. If we'd run with docker, we need to persist the storage anyway and might as well persist the configuration file. I am not sure if scraping those metrics is worth the extra layer, but it should be no difference to launch the binary directly or launching it dockerified mounting the same persistent storage to the container.

Reply:
Yep, we mount a host directory in the containers so the repo is persistent across restarts and image updates. 
We're already running go-ipfs (alongside nginx, node_exporter, etc) in Docker containers on the bootstrappers and have no plans to switch to systemd. I think it would be very beneficial to standardize the way we do deployments across gateways,bootstrappers,preloaders and Docker has no particular drawbacks that I am aware of. 
What in particular do you think is a major concern for running go-ipfs in Docker ?

Reply:
> Yep, we mount a host directory in the containers so the repo is persistent across restarts and image updates.

Then we don't need to set all the env vars with special configs for docker?

> We're already running go-ipfs (alongside nginx, node_exporter, etc) in Docker containers on the bootstrappers

I will try to figure out why bootstrappers need nginx..

> What in particular do you think is a major concern for running go-ipfs in Docker ?

Major or minor, but:
- Custom docker build and special configuration: more code and redirection, more places to mess up something
- Docker itself: one more layer of indirection, history of horrible bugs, perf impact

The strongest reason for a docker approach I see above is about the individual container metrics. Since we are not using those now we'll have to see if it's something we need. I'm not saying no to docker, but it we don't have an actual need for it I'd rather not run it. And if we do, we should pick up the upstream images (if we have improvements to the builds, they should be PRed upstream).

Reply:
> I will try to figure out why bootstrappers need nginx..

Bootstrappers used to be gateways at the same time.

Reply:
I think it is to put ssl and basic auth in front of node_exporter

Reply:
NGINX on bootstrappers exposes the go-ipfs' WebSocket port

Reply:
Also, disagree on metrics. Just because we're not collecting them now doesn't mean that we shouldn't, they provide very valuable insight into go-ipfs' usage profile and make it easy to follow evolution over time.

Reply:
Just because we aren't collecting the metrics right now means we should start collecting them not remove our ability to collect them.